### PR TITLE
Path starts (not) with

### LINKED ISSUE #187: Path.startsWith
*Platform* (jvm, jdk8, js, android): jdk8

## Code related feature
```
expect(path).startsWith(otherPath)

//instead of

expect(path).feature { f1(it::startsWith, otherpath) }.toBe(true) 
```
However, in reporting we only want to see (in contrast to the `feature` version):
```
expect: Path...
* starts with: Path
```

Following the things you need to do:

*domain*
- extend PathAssertions with a function `startsWith` (see other functions as a guideline)
- modify PathAssertionsBuilder, delegate to pathAssertions (see other functions as a guideline)
- delegate implementation to robstoll-lib in PathAssertionsImpl (see other functions as a guideline)

*lib*
- implement _startsWith in pathAssertions by using the `ExpectImpl.builder.descriptive...` (see /domain/robstoll-lib/atrium-domain-robstoll-lib-common/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/charSequenceAssertions.kt -> _startsWith)
  - add `STARTS_WITH("starts with")` to DescriptionPathAssertion for en_GB
  - add `STARTS_WITH("beginnt mit"` to DescriptionPathAssertion for de_CH

*api* (fluent-en_GB/extension/jdk8)
- provide a fun `startsWith`
- add `@since 0.9.0` to KDOC
- add `startsWith` to PathAssertionsSpec in specs-common and extend it in atrium-api-fluent-en_GB-common/src/test

## Your first contribution?
- Write a comment `I'll work on this` if you would like to take this issue over. 
  This way we get the chance to revise the description in case things have changed in the meantime,
  we might give you additional hints and we can assign the task to you, so that others do not start as well.
- See [Your first code contribution](https://github.com/robstoll/atrium/blob/master/.github/CONTRIBUTING.md#your-first-code-contribution) for guidelines.  
- Do not hesitate to ask questions here or to contact us via [Atrium's slack channel](https://kotlinlang.slack.com/team/U3DE1TXKP) if you need help
  ([Invite yourself](https://slack.kotlinlang.org/) in case you do not have an account yet).

### LINKED ISSUE #189: Path.startsNotWith
*Platform* (jvm, jdk8, js, android): jdk8

## Code related feature
```
expect(path).startsNotWith(otherPath)

//instead of

expect(path).feature { f1(it::startsNotWith, otherpath) }.toBe(true) 
```
However, in reporting we only want to see (in contrast to the `feature` version):
```
expect: Path...
* starts not with: Path
```

Following the things you need to do:

*domain*
- extend PathAssertions with a function `startsNotWith` (see other functions as a guideline)
- modify PathAssertionsBuilder, delegate to pathAssertions (see other functions as a guideline)
- delegate implementation to robstoll-lib in PathAssertionsImpl (see other functions as a guideline)

*lib*
- implement _startsWith in pathAssertions by using the `ExpectImpl.builder.descriptive...` (see /domain/robstoll-lib/atrium-domain-robstoll-lib-common/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/charSequenceAssertions.kt -> _startsNotWith)
  - add `STARTS_NOT_WITH("starts not with")` to DescriptionPathAssertion for en_GB
  - add `STARTS_NOT_WITH("beginnt nicht mit"` to DescriptionPathAssertion for de_CH

*api* (fluent-en_GB/extension/jdk8)
- provide a fun `startsWith`
- add `@since 0.9.0` to KDOC
- add `startsNotWith` to PathAssertionsSpec in specs-common and extend it in atrium-api-fluent-en_GB-common/src/test

## Your first contribution?
- Write a comment `I'll work on this` if you would like to take this issue over. 
  This way we get the chance to revise the description in case things have changed in the meantime,
  we might give you additional hints and we can assign the task to you, so that others do not start as well.
- See [Your first code contribution](https://github.com/robstoll/atrium/blob/master/.github/CONTRIBUTING.md#your-first-code-contribution) for guidelines.  
- Do not hesitate to ask questions here or to contact us via [Atrium's slack channel](https://kotlinlang.slack.com/team/U3DE1TXKP) if you need help
  ([Invite yourself](https://slack.kotlinlang.org/) in case you do not have an account yet).

### PR DESCRIPTION
**Pull request by arjank**
Add `path starts with` and `path starts not with` assertions, including translations and some tests to show functionality.

Closes #187 
Closes #189 

----
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/master/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

File: apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/PathAssertionsSpec.kt [hunk 1]
```diff
@@ -9,6 +9,8 @@ import java.nio.file.Path
 class PathAssertionsSpec : ch.tutteli.atrium.specs.integration.PathAssertionsSpec(
     fun0(Expect<Path>::exists),
     fun0(Expect<Path>::existsNot),
+    fun1(Expect<Path>::startsWith),
+    fun1(Expect<Path>::startsNotWith),
     fun1(Expect<Path>::endsWith)
 ){
     @Suppress("unused", "UNUSED_VALUE")
```
Comment (robstoll) [inline apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/PathAssertionsSpec.kt]: please add startsWith/startsNotWith also to ambiguityTest (and please rebase your branch on master)

File: misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/PathAssertionsSpec.kt [hunk 2]
```diff
@@ -91,6 +93,30 @@ abstract class PathAssertionsSpec(
         }
     }
 
+    describeFun(startsWith.name) {
+        val startsWithFun = startsWith.lambda
+        context("starts with") {
+            it("does not throw") {
```
Comment (robstoll) [inline misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/PathAssertionsSpec.kt]: please also include a use case which fails (see spec for endsWith)

File: misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/PathAssertionsSpec.kt [hunk 3]
```diff
@@ -91,6 +93,30 @@ abstract class PathAssertionsSpec(
...
+
+    describeFun(startsNotWith.name) {
+        val startsNotWithFun = startsNotWith.lambda
+        context("starts not with") {
+            it("does not throw") {
+                expect(Paths.get("/some/path/for/test"))
+                    .startsNotWithFun(Paths.get("/other/path/"))
+            }
+            it("does not match partials") {
+                expect(Paths.get("/some/path/for/test"))
```
Comment (robstoll) [inline misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/PathAssertionsSpec.kt]: same here, use case which fails

File: apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/PathAssertionsSpec.kt (hunk 1 above)
Comment (arjank) [reply apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/PathAssertionsSpec.kt]: Requested tests added and also rebased.

**Comment by arjank** [comment]
Workflow build failed, seems Github related. I'm not sure how we can trigger this again (without adding new commits).

File: apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/pathAssertions.kt [hunk 4]
```diff
@@ -5,6 +5,28 @@ import ch.tutteli.atrium.domain.builders.ExpectImpl
 import ch.tutteli.atrium.domain.builders.path
 import java.nio.file.Path
 
+/**
+ * Expects that the subject of the assertion (a [Path]) starts with the expected [Path].
```
Comment (robstoll) [inline apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/pathAssertions.kt]: I tried to retrigger github action but without success at it seems. Maybe a new commit helps. Thus a small improvement (same can be done for the other kdoc)
```suggestion
 * Expects that the subject of the assertion (a [Path]) starts with the [expected] [Path].
```

File: apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/pathAssertions.kt [hunk 5]
```diff
@@ -5,6 +5,28 @@ import ch.tutteli.atrium.domain.builders.ExpectImpl
...
+ * @return This assertion container to support a fluent API.
+ * @throws AssertionError Might throw an [AssertionError] if the assertion made is not correct.
+ *
+ * @since 0.9.0
+ */
+fun <T : Path> Expect<T>.startsWith(expected: Path): Expect<T> =
+    addAssertion(ExpectImpl.path.startsWith(this, expected))
+
+/**
+ * Expects that the subject of the assertion (a [Path]) does not start with the expected [Path].
```
Comment (robstoll) [inline apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/pathAssertions.kt]:
```suggestion
 * Expects that the subject of the assertion (a [Path]) does not start with the [expected] [Path].
```

File: misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/PathAssertionsSpec.kt [hunk 6]
```diff
@@ -91,6 +93,56 @@ abstract class PathAssertionsSpec(
...
+            it("does not throw") {
+                expect(Paths.get("/some/path/for/test"))
+                    .startsWithFun(Paths.get("/some/path/"))
+            }
+        }
+
+        val expectedMessageIfNotStartsWith = "${DescriptionPathAssertion.STARTS_WITH.getDefault()}:"
+
+        context("does not start with") {
+            it("throws an AssertionException") {
```
Comment (robstoll) [inline misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/PathAssertionsSpec.kt]:
```suggestion
            it("throws an AssertionError") {
```

File: misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/PathAssertionsSpec.kt [hunk 7]
```diff
@@ -91,6 +93,56 @@ abstract class PathAssertionsSpec(
...
+            it("does not match partials") {
+                expect(Paths.get("/some/path/for/test"))
+                    .startsNotWithFun(Paths.get("/some/pa"))
+            }
+        }
+
+        val expectedMessageIfStartsWith = "${DescriptionPathAssertion.STARTS_NOT_WITH.getDefault()}:"
+
+        context("starts with") {
+            it("throws an AssertionException") {
```
Comment (robstoll) [inline misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/PathAssertionsSpec.kt]:
```suggestion
            it("throws an AssertionError") {
```

**Comment by robstoll** [comment]
@arjank thanks for the contribution :+1:

Fancy another one:
https://github.com/robstoll/atrium/issues?q=is%3Aissue+is%3Aopen+label%3Ahacktoberfest+no%3Aassignee

**Pull request merged**